http/src/main/java/io/micronaut/http/body/DynamicWriter.java

```diff
@@ -0,0 +1,60 @@
+package io.micronaut.http.body;
+
+import io.micronaut.core.annotation.Internal;
+import io.micronaut.core.io.buffer.ByteBuffer;
+import io.micronaut.core.io.buffer.ByteBufferFactory;
+import io.micronaut.core.reflect.ClassUtils;
+import io.micronaut.core.type.Argument;
+import io.micronaut.core.type.MutableHeaders;
+import io.micronaut.http.MediaType;
+import io.micronaut.http.codec.CodecException;
+
+import java.io.OutputStream;
+import java.util.List;
+import java.util.Optional;
+
+@Internal
+public final class DynamicWriter implements MessageBodyWriter<Object> {
+    private final MessageBodyHandlerRegistry registry;
+    private final List<MediaType> mediaTypes;
+
+    public DynamicWriter(MessageBodyHandlerRegistry registry, List<MediaType> mediaTypes) {
+        this.registry = registry;
+        this.mediaTypes = mediaTypes;
+    }
+
+    @Override
+    public MessageBodyWriter<Object> createSpecific(Argument<Object> type) {
+        return registry.findWriter(type, mediaTypes).orElse(this);
+    }
+
+    public MessageBodyWriter<Object> find(Argument<Object> type, MediaType mediaType, Object object) {
+        Optional<MessageBodyWriter<Object>> specific = registry.findWriter(type, List.of(mediaType));
+        if (specific.isPresent() && !(specific.get() instanceof DynamicWriter)) {
```

Comment:
If the writer is shared maybe we can store `lastFound` and `lastArgument` if it matches, return it, avoiding the writer.

Reply:
DynamicWriter is not shared. It used to be (the registry used to return it as fallback) but it didn't turn out to work so it's used directly now.